core/src/main/java/org/apache/carbondata/core/scan/executor/util/RestructureUtil.java

```diff
@@ -167,15 +167,22 @@ public static boolean isColumnMatches(boolean isTransactionalTable,
     // column ID but can have same column name
     if (tableColumn.getDataType().isComplexType() && !(tableColumn.getDataType().getId()
         == DataTypes.ARRAY_TYPE_ID)) {
-      if (tableColumn.getColumnId().equalsIgnoreCase(queryColumn.getColumnId())) {
+      if (tableColumn.getColumnId().equalsIgnoreCase(queryColumn.getColumnId()) || (
```

Comment:
Can you override equals method of Column or add method inside a column to do this check. It is pretty repetitive and more prone to issues.